transcendence-app/migrations/20221201150119_add_localfile_table.ts

```diff
@@ -0,0 +1,21 @@
+import { Knex } from 'knex';
+
+export async function up(knex: Knex): Promise<void> {
```

Comment:
Aside from filling up our tables on startup, could you elaborate what other benefits we gain? (new for me and wish to learn :p )

Reply:
Currently, the database initialization scripts are only run if we start the container with an empty data directory.
And when we want to edit the database schema, we delete the volumes and restart the containers, which run the initialization scripts.

Or, if we want preserve the data in our database, we could make schema changes manually without deleting the volumes.

With migrations, we can update the database schema automatically without nuking the database or doing migrations manually.

If we want to add a new table, update the column or a table, or make any schema change, we can create a migration that has `up` and `down` functions, and we can run the following CLI commands:
```shell
knex migrate:latest # run all the migrations from the current version until the latest one
knex migrate:rollback # rollback to the previous version
knex migrate:up # run one migration, from the current version to the next version
knex migrate:down # rollback one migration from the current version to the previous version
```

Reply:
As soon as you do a down, you nuke the table, at least in the current implementation. I would expect a system to map items from a previous table scheme to a new table scheme, that would run for each of the items in one of the tables, for each of the tables.
How do you do that? What am I missing?

Reply:
I have another question, this means with migrations we would be able to deploy new versions without affecting already existing data in the DB? (even if we make an schema upgrade)

Reply:
> As soon as you do a down, you nuke the table, at least in the current implementation. I would expect something like a system to map items from a previous table scheme to a new table scheme, that would run for each of the items in one of the tables, for each of the tables, that saves the old data in memory while it does the swap, and then removes the old invalid data.
> How do you do that? What am I missing?

I don't understand this question. Could you clarify, please?
The current migrations create tables. The opposite of creating a table is deleting the table (drop table).

Reply:
> I have another question, this means with migrations we would be able to deploy new versions without affecting already existing data in the DB? (even if we make an schema upgrade)

Yes, we would be able to do that.